SSHClient: request host key type from known_hosts

ploxiln opened the pull request:
fixes #865 

one part of an alternative set of PRs to #899 

I only backported this to 2.1 because there were some changes around `t.start_client()` in 2.0 and 2.1 which make forward-merging have possibly tricky conflicts there. I can backport further if you would like, and won't mind the conflict resolution when forward-merging :)

bitprophet commented:
Nah, at this point anything that makes backporting to 1.x is a good argument for "don't bother".

If someone who's stuck on 1.x (despite the security concerns, which was one reason we even did the backend switch in 2.0) is super upset they don't have such a change, they can submit their own PR :D

ploxiln commented:
I can rebase this and resolve conflicts if you like (... #974 causes conflicts for pretty much all pull requests against 2.0+ ...)

bitprophet commented:
That'd be great, thanks! (And, yes, it does, but I'd rather not let compatibility with PRs prevent necessary change. For example all of Fabric 1's PRs will get broked good once 2.0 becomes the new master... 🙃

ploxiln commented:
Rebased, resolved conflicts, and also removed a commit which tweaked a log message which I realized is not related to host keys (it's related to user auth keys).

Manually tested and verified, but I'm working on a proper automated test...

ploxiln commented:
Added proper tests for this. If I run them against the 2.1 branch (without this PR), the ones which I expect to fail do so:

```
test_host_key_negotiation_1 (test_client.SSHClientTest) ... /Users/pierce/scratch/fabwork/paramiko/paramiko/client.py:708: UserWarning: Unknown ssh-rsa host key for [127.0.0.1]:50456: 60733844cb5186657fdedaa22b5a57d5
  key.get_fingerprint())))
FAIL
test_host_key_negotiation_2 (test_client.SSHClientTest) ... ok
test_host_key_negotiation_3 (test_client.SSHClientTest) ... ERROR
test_host_key_negotiation_4 (test_client.SSHClientTest) ... ok
```

ploxiln commented:
Now that #981 is merged, I can revise this to use the SecurityOptions api to change the order of the preferred host key types (to put the available key first). I may do very soon that if it's not too late...

ploxiln commented:
changed to use SecurityOptions, will squash down a bit if that looks good

bitprophet commented:
Hrm, this appears to conflict with the ResourceManager removal due to the same chunk moving around, but I think it's just re-applying 2 line nukes after merging your changes, so I've done that.

ploxiln commented:
Your merge conflict resolution looks correct - thanks!
